File: lib/models/order.dart
import 'package:restaurant_app/models/cart_item.dart';



class Order {
  final String id;
  final String userId;
  final String customerName;
  final List<CartItem> items;
  final double totalPrice;
  String status; // Pending, Approved, Delivered, Cancelled
  final DateTime date;

  Order({
    required this.id,
     required this.userId,
    required this.customerName,
    required this.items,
    required this.totalPrice,
    required this.status,
    required this.date,
  });
    factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json["_id"] ?? "",
      userId: json["userId"] ?? "",
      customerName: json["customerName"] ?? "",
      items: (json["items"] as List? ?? [])
          .map((i) => CartItem.fromJson(i))
          .toList(),
      totalPrice: (json["totalPrice"] ?? 0).toDouble(),
      status: json["status"] ?? "Pending",
      date: DateTime.tryParse(json["date"] ?? "") ?? DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "customerName": customerName,
      "items": items.map((i) => i.toJson()).toList(),
      "totalPrice": totalPrice,
      "status": status,
      "date": date.toIso8601String(),
    };
  }

}

File: backend/routes/order.js
const router = require("express").Router();
const Order = require("../models/Order");
const verifyToken = require("../middleware/verifyToken");
const verifyAdmin = require("../middleware/verifyAdmin");



// CREATE PAYMENT INTENT (AUTH)
router.post("/payment-intent", verifyToken, async (req, res) => {
  try {
    
    const amount = Number(req.body.amount);
    const currency = (req.body.currency || "usd").toString().toLowerCase();

    if (!amount || amount <= 0 || !Number.isInteger(amount)) {
      return res
        .status(400)
        .json({ message: "amount must be a positive integer in cents" });
    }

    const paymentIntent = await stripe.paymentIntents.create({
      amount,
      currency,
      automatic_payment_methods: { enabled: true },
      metadata: {
        userId: req.user.id,
      },
    });

    return res.status(200).json({
      clientSecret: paymentIntent.client_secret,
    });
  } catch (err) {
    return res.status(500).json(err);
  }
});

// CREATE ORDER (AUTH)
router.post("/", verifyToken, async (req, res) => {
  try {
    const order = new Order({
      userId: req.user.id,
      customerName: req.body.customerName,
      items: req.body.items,
      totalPrice: req.body.totalPrice,
      status: req.body.status || "Pending",
      date: req.body.date || Date.now(),
    });

    const saved = await order.save();
    res.status(201).json(saved);
  } catch (err) {
    res.status(500).json(err);
  }
});

// GET MY ORDERS (AUTH)
router.get("/my", verifyToken, async (req, res) => {
  try {
    const orders = await Order.find({ userId: req.user.id }).sort({ createdAt: -1 });
    res.status(200).json(orders);
  } catch (err) {
    res.status(500).json(err);
  }
});

// GET ALL ORDERS (ADMIN)
router.get("/", verifyToken, verifyAdmin, async (req, res) => {
  try {
    const orders = await Order.find().sort({ createdAt: -1 });
    res.status(200).json(orders);
  } catch (err) {
    res.status(500).json(err);
  }
});

// UPDATE STATUS (ADMIN)
router.put("/:id/status", verifyToken, verifyAdmin, async (req, res) => {
  try {
    const updated = await Order.findByIdAndUpdate(
      req.params.id,
      { $set: { status: req.body.status } },
      { new: true }
    );
    res.status(200).json(updated);
  } catch (err) {
    res.status(500).json(err);
  }
});

module.exports = router;
